monitoring/perf_context_imp.h

```diff
@@ -37,57 +42,62 @@ extern thread_local PerfContext perf_context;
   perf_step_timer_##metric.Start();
 
-#define PERF_CONDITIONAL_TIMER_FOR_MUTEX_GUARD(metric, condition, stats,       \
-                                               ticker_type)                    \
-  PerfStepTimer perf_step_timer_##metric(&(perf_context.metric), nullptr,      \
-                                         false, PerfLevel::kEnableTime, stats, \
-                                         ticker_type);                         \
-  if (condition) {                                                             \
-    perf_step_timer_##metric.Start();                                          \
+#define PERF_CONDITIONAL_TIMER_FOR_MUTEX_GUARD(metric, condition, stats, \
+                                               ticker_type)              \
+  PerfStepTimer perf_step_timer_##metric(                                \
+      &(perf_context.metric), nullptr, false,                            \
+      (bool)perf_flags.enable_##metric##_bit, stats, ticker_type);       \
+  if (condition) {                                                       \
+    perf_step_timer_##metric.Start();                                    \
   }
 
 // Update metric with time elapsed since last START. start time is reset
 // to current timestamp.
 #define PERF_TIMER_MEASURE(metric) perf_step_timer_##metric.Measure();
 
 // Increase metric value
-#define PERF_COUNTER_ADD(metric, value)        \
-  if (perf_level >= PerfLevel::kEnableCount) { \
-    perf_context.metric += value;              \
+#define PERF_COUNTER_ADD(metric, value)   \
+  if (perf_flags.enable_##metric##_bit) { \
+    perf_context.metric += value;         \
   }
 
 // Increase metric value
-#define PERF_COUNTER_BY_LEVEL_ADD(metric, value, level)                      \
-  if (perf_level >= PerfLevel::kEnableCount &&                               \
-      perf_context.per_level_perf_context_enabled &&                         \
-      perf_context.level_to_perf_context) {                                  \
-    if ((*(perf_context.level_to_perf_context)).find(level) !=               \
-        (*(perf_context.level_to_perf_context)).end()) {                     \
-      (*(perf_context.level_to_perf_context))[level].metric += value;        \
-    }                                                                        \
-    else {                                                                   \
-      PerfContextByLevel empty_context;                                      \
-      (*(perf_context.level_to_perf_context))[level] = empty_context;        \
-      (*(perf_context.level_to_perf_context))[level].metric += value;       \
-    }                                                                        \
-  }                                                                          \
+// TODO: add a flag count
+#define PERF_COUNTER_BY_LEVEL_ADD(metric, value, level)               \
+  if (perf_context.per_level_perf_context_enabled &&                  \
```

Comment:
Seems that there is an optimization in the original implementation that it checks whether the perf is enabled. You should check it as well I think, considering that checking perf flags should be faster than the `xxx.find()` below.